### PR TITLE
Adds App store endpoints

### PR DESCRIPTION
**Pull request by SerKnight**
Implements the endpoints available in Xero's API spec.

## Description

https://developer.xero.com/documentation/guides/how-to-guides/xero-app-store-subscriptions/

https://developer.xero.com/documentation/api/xero-app-store/subscriptions

## Release Notes
Subscriptions endpoints enable you to query data about the users who have paid to subscribe to your application through Xero.

AppStoreAPI adds
*  `getSubscription` 

AppStore adds
* Subscription
* SubscriptionItem
* Plan
* Product
* Price
* ProblemDetails

---
Employee adds `terminationReason`
`V` Voluntary cessation
`I` Ill health
`D` Deceased
`R` Redundancy
`F` Dismissal
`C` Contract cessation
`T` Transfer

* PayrollCalendar adds `reference_date`

* Payment.rb adds `bank_amount`

## Types of Changes
- [x] New feature (non-breaking change that adds functionality)

### CODE REVIEW

File: src/XeroClient.ts [hunk 1]
```diff
@@ -185,8 +187,8 @@ export class XeroClient {
   }
 
   public async getClientCredentialsToken(): Promise<TokenSet> {
-    const { clientId, clientSecret, grantType } = this.config;
-    const result = await this.tokenRequest(clientId, clientSecret, { grant_type: grantType });
+    const { clientId, clientSecret, grantType, scopes } = this.config;
```
Comment (RettBehrens) [inline src/XeroClient.ts:190]: It shouldn't be an issue but might be worth testing this change with the custom connections starter app to confirm `scopes` being `undefined` doesn't break anything

File: src/gen/api/appStoreApi.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
+/**
+ * Xero AppStore API
+ * These endpoints are for Xero Partners to interact with the App Store Billing platform
+ *
+ * The version of the OpenAPI document: 2.12.0
```
Comment (RettBehrens) [inline src/gen/api/appStoreApi.ts:5]: should this be 2.15.0?

File: src/gen/api/assetApi.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@
  * Xero Assets API
  * The Assets API exposes fixed asset related functions of the Xero Accounting application and can be used for a variety of purposes such as creating assets, retrieving asset valuations etc.
  *
- * The version of the OpenAPI document: 2.13.2
+ * The version of the OpenAPI document: 2.13.5
```
Comment (RettBehrens) [inline src/gen/api/assetApi.ts:5]: should this be 2.15.0?

File: src/gen/api/bankfeedsApi.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@
  * Xero Bank Feeds API
  * The Bank Feeds API is a closed API that is only available to financial institutions that have an established financial services partnership with Xero. If you\'re an existing financial services partner that wants access, contact your local Partner Manager. If you\'re a financial institution who wants to provide bank feeds to your business customers, contact us to become a financial services partner.
  *
- * The version of the OpenAPI document: 2.13.2
+ * The version of the OpenAPI document: 2.13.5
```
Comment (RettBehrens) [inline src/gen/api/bankfeedsApi.ts:5]: should this be 2.15.0?

File: src/gen/api/payrollAUApi.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@
  * Xero Payroll AU API
  * This is the Xero Payroll API for orgs in Australia region.
  *
- * The version of the OpenAPI document: 2.13.2
+ * The version of the OpenAPI document: 2.13.5
```
Comment (RettBehrens) [inline src/gen/api/payrollAUApi.ts:5]: should this be 2.15.0?

File: src/gen/api/payrollNZApi.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@
  * Xero Payroll NZ
  * This is the Xero Payroll API for orgs in the NZ region.
  *
- * The version of the OpenAPI document: 2.13.2
+ * The version of the OpenAPI document: 2.13.5
```
Comment (RettBehrens) [inline src/gen/api/payrollNZApi.ts:5]: should this be 2.15.0?

File: src/gen/api/payrollUKApi.ts [hunk 7]
```diff
@@ -2,7 +2,7 @@
  * Xero Payroll UK
  * This is the Xero Payroll API for orgs in the UK region.
  *
- * The version of the OpenAPI document: 2.13.2
+ * The version of the OpenAPI document: 2.13.5
```
Comment (RettBehrens) [inline src/gen/api/payrollUKApi.ts:5]: should this be 2.15.0?

File: src/gen/api/projectApi.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@
  * Xero Projects API
  * This is the Xero Projects API
  *
- * The version of the OpenAPI document: 2.13.2
+ * The version of the OpenAPI document: 2.13.5
```
Comment (RettBehrens) [inline src/gen/api/projectApi.ts:5]: should this be 2.15.0?

File: src/test/xeroClient.spec.ts [hunk 9]
```diff
@@ -100,6 +100,7 @@ describe('the XeroClient', () => {
       expect(xeroClient).toHaveProperty('payrollAUApi')
       expect(xeroClient).toHaveProperty('payrollUKApi')
       expect(xeroClient).toHaveProperty('payrollNZApi')
+      expect(xeroClient).toHaveProperty('appStoreApi')
```
Comment (RettBehrens) [inline src/test/xeroClient.spec.ts:103]: Are there any other low hanging fruit / sanity check tests we could add around the app store API test coverage?

**Review by RettBehrens — CHANGES_REQUESTED** [review]
Please see comments before merging

File: README.md [hunk 10]
```diff
@@ -148,6 +150,77 @@ Because Custom Connections are only valid for a single organisation you don't ne
...
+  testMode: true
+}
+```
+You should use the subscription data to provision user access/permissions to your application.
+### App Store Subscription Webhooks
+
+In additon to a subscription Id being passed through the URL, when a purchase or an upgrade takes place you will be notified via a webhook. You can then use the subscription Id in the webhook payload to query the AppStore endpoints and determine what plan the user purchased, upgraded, downgraded or cancelled.
+
+Refer to Xero's documenation to learn more about setting up and receiving webhooks.
+> https://developer.xero.com/documentation/guides/webhooks/overview/
```
[the remark this answers is not in the document]
Comment (SerKnight) [reply README.md:222]: Added!

File: src/XeroClient.ts (hunk 1 above)
Comment (SerKnight) [reply src/XeroClient.ts:190]: 👍  Tested in custom connections app w/ out scopes for a normal Custom Connections app

Screen Shot 2021-08-05 at 1 49 51 PM
Screen Shot 2021-08-05 at 1 51 22 PM

File: src/gen/api/appStoreApi.ts (hunk 2 above)
Comment (SerKnight) [reply src/gen/api/appStoreApi.ts:5]: Yeah I will regen once I've got master merged for OAS pr which will update this 

https://github.com/XeroAPI/Xero-OpenAPI/pull/440/files

File: src/test/xeroClient.spec.ts (hunk 9 above)
Comment (SerKnight) [reply src/test/xeroClient.spec.ts:103]: This is the lowest of fruit haha. There is really only 1 new function being added. Setting up http://getsandbox.com/ for node sdk would be a great addition though

**Review by RettBehrens — APPROVED** [review]

**Pull request merged**